dart/common/Extensible.h

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef DART_COMMON_EXTENSIBLE_H_
+#define DART_COMMON_EXTENSIBLE_H_
+
+#include <memory>
+#include <vector>
+
+namespace dart {
+namespace common {
+
+/// Extensible is a CRTP base class that provides an interface for easily
+/// creating data structures that are meant to be extended. Ordinary copying
+/// does not work with extended data structures, because information from the
+/// derived classes will typically be lost during a copy. The Extensible class
+/// provides an interface for creating a new copy of an extended data structure,
+/// as well as copying information between two extended data structures of the
+/// same type.
+template <class T>
+class Extensible
+{
+public:
+
+  /// Default constructor
+  Extensible() = default;
+
+  /// Virtual destructor
+  virtual ~Extensible() = default;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible(const Extensible& doNotCopy) = delete;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible& operator=(const Extensible& doNotCopy) = delete;
```

Comment:
``Extensible`` gets used as a foundation in a higher level framework. As an example, ``Addon::State`` is an ``Extensible``, and so ``AddonManager::State`` is an ``ExtensibleMapHolder`` for an ``AddonManager::StateMap``.

The ``AddonManager::StateMap`` is the key to keeping everything sane. ``AddonManager::StateMap`` is a typedef for ``std::map< std::type_index, std::unique_ptr<Addon::State> >``. This means that ``AddonManager::StateMap`` is mapping a set of ``Addon::State`` pointers to the ``Addon`` types that they correspond to.

``ExtensibleMapHolder`` then wraps up the ``Addon::StateMap`` to (1) obey the rule of five for deep copying and (2) be more efficient about copying, as discussed in another thread.

So the basic pipeline is:

1. Get an ``AddonManager::State`` object by calling ``AddonManager::getAddonStates()``.
2. Store the ``AddonManager::State`` object however you please, since it follows the rule of five (**but** you can't manipulate its contents, because the entries are protected).
3. Pass the ``AddonManager::State`` object back into the ``AddonManager`` it came from, or into any other ``AddonManager`` (any mismatches in the Addons will be handled by ignoring them).

So the users should never actually be touching raw ``Extensible`` objects themselves. Instead, they should just be passing around these higher level wrapper classes, like ``AddonManager::State``.